File: python_programs_and_containers/building_blocks/knowledge_base/kb_python/postgres/data_structures/kb_status_data.py
from typing import Dict, Any, Tuple,Union,List
from psycopg2 import sql, OperationalError, InterfaceError
import json
import time

class KB_Status_Data:
    """
    A class to handle the status data for the knowledge base.
    Performance-optimized with proper error handling and dynamic table names.
    """
    
    def __init__(self, kb_search, database):
        """
        Initialize the KB_Status_Data object.
        
        Args:
            kb_search: An instance of KB_Search class
            database (str): The base database name
        """
        self.kb_search = kb_search
        self.base_table = f"{database}_status"
    
    def find_node_id(self, kb, node_name, properties=None, node_path=None):
        """
        Find a single node id for given parameters. Raises error if 0 or multiple nodes found.
        
        Args:
            kb (str): Knowledge base name
            
            node_name (str): Node name to search for
            properties (dict, optional): Properties to match
            node_path (str, optional): LTREE path to match
            
        Returns:
            dict: Single matching node record
            
        Raises:
            ValueError: If no node or multiple nodes found
        """
        
        
        results = self.find_node_ids(kb, node_name, properties, node_path)
        
        if len(results) == 0:
            raise ValueError(f"No node found matching parameters: kb={kb}, name={node_name}, properties={properties}, path={node_path}")
        if len(results) > 1:
            raise ValueError(f"Multiple nodes ({len(results)}) found matching parameters: kb={kb}, name={node_name}, properties={properties}, path={node_path}")
        
        return results[0]
    
    def find_node_ids(self, kb=None,  node_name=None, properties=None, node_path=None):
        """
        Find all node ids matching the given parameters.
        
        Args:
            kb (str, optional): Knowledge base name
            node_name (str, optional): Node name to search for
            properties (dict, optional): Properties to match
            node_path (str, optional): LTREE path to match
            
        Returns:
            list: List of matching node records
            
        Raises:
            ValueError: If no nodes found
        """
        
        
        try:
            # Clear previous filters and build new query
            self.kb_search.clear_filters()
            self.kb_search.search_label("KB_STATUS_FIELD")
            
            if kb is not None:
                self.kb_search.search_kb(kb)
          
            if node_name is not None:
                self.kb_search.search_name(node_name)
            if properties is not None and isinstance(properties, dict):
                for key, value in properties.items():
                    self.kb_search.search_property_value(key, value)
            if node_path is not None:
                self.kb_search.search_path(node_path)
            
            # Execute query and get results
            node_ids = self.kb_search.execute_query()
            
            if not node_ids or len(node_ids) == 0:
                raise ValueError(f"No nodes found matching parameters: kb={kb}, name={node_name}, properties={properties}, path={node_path}")
            
            return node_ids
            
        except Exception as e:
            if isinstance(e, ValueError):
                raise
            raise Exception(f"Error finding node IDs: {str(e)}")
    
 
    def get_status_data(self, path):
        """
        Retrieve status data for a given path with performance optimization.
        
        Args:
            path (str): The path to search for
            
        Returns:
            tuple: (data_dict, path) where data_dict is a Python dictionary
            
        Raises:
            Exception: If no data is found or a database error occurs
        """
        if not path:
            raise ValueError("Path cannot be empty or None")
        
        try:
            # Use parameterized query with the dynamic table name
            query = f"""
            SELECT data, path
            FROM {self.base_table}
            WHERE path = %s
            LIMIT 1
            """
            
            self.kb_search.cursor.execute(query, (path,))
            result = self.kb_search.cursor.fetchone()
            
            if result is None:
                raise Exception(f"No data found for path: {path}")
            
            # Extract data and path from result (handle both dict and tuple)
            if isinstance(result, dict):
                data = result['data']
                path_value = result['path']
            else:
                data, path_value = result
            
            # Handle JSON conversion if needed
            if isinstance(data, str):
                try:
                    data = json.loads(data)
                except json.JSONDecodeError as e:
                    raise Exception(f"Failed to decode JSON data for path '{path}': {str(e)}")
            
            return data, path_value
            
        except Exception as e:
            if str(e).startswith(("No data found", "Failed to decode", "Path cannot be")):
                raise
            raise Exception(f"Error retrieving status data for path '{path}': {str(e)}")
    
    def get_multiple_status_data(self, paths):
        """
        Performance-optimized method to retrieve status data for multiple paths in a single query.
        
        Args:
            paths (list): List of paths to search for
            
        Returns:
            dict: Dictionary mapping paths to their data
        """
        if not paths:
            return {}
        
        if not isinstance(paths, (list, tuple)):
            paths = [paths]
        
        try:
            # Use IN clause for efficient bulk retrieval
            placeholders = ','.join(['%s'] * len(paths))
            query = f"""
            SELECT data, path
            FROM {self.base_table}
            WHERE path IN ({placeholders})
            """
            
            self.kb_search.cursor.execute(query, tuple(paths))
            results = self.kb_search.cursor.fetchall()
            
            data_dict = {}
            for result in results:
                if isinstance(result, dict):
                    data = result['data']
                    path_value = result['path']
                else:
                    data, path_value = result
                
                # Handle JSON conversion if needed
                if isinstance(data, str):
                    try:
                        data = json.loads(data)
                    except json.JSONDecodeError:
                        # Log warning but continue with string data
                        print(f"Warning: Failed to decode JSON for path '{path_value}'")
                
                data_dict[path_value] = data
            
            return data_dict
            
        except Exception as e:
            raise Exception(f"Error retrieving multiple status data: {str(e)}")


    def set_status_data(self, path: str, data: Dict[str, Any], 
                    retry_count: int = 3, retry_delay: float = 1.0) -> Tuple[bool, str]:
        """
        Update status data for a given path with performance optimization using UPSERT.
        
        Args:
            path (str): The path to update
            data (dict): The data to store
            retry_count (int): Number of retries on failure (default: 3)
            retry_delay (float): Delay between retries in seconds (default: 1.0)
            
        Returns:
            tuple: (bool, str) indicating success/failure and a message
            
        Raises:
            ValueError: If inputs are invalid
            Exception: If the update fails after all retries
        """
        # Input validation
        if not path:
            raise ValueError("Path cannot be empty or None")
        if not isinstance(data, dict):
            raise ValueError("Data must be a dictionary")
        if retry_count < 0:
            raise ValueError("Retry count must be non-negative")
        if retry_delay < 0:
            raise ValueError("Retry delay must be non-negative")
        
        # Convert data to JSON string once, outside the retry loop
        json_data = json.dumps(data, separators=(',', ':'))
        
        # Prepare the query
        upsert_query = sql.SQL("""
            INSERT INTO {table} (path, data)
            VALUES (%s, %s)
            ON CONFLICT (path)
            DO UPDATE SET data = EXCLUDED.data
            RETURNING path, (xmax = 0) AS was_inserted
        """).format(
            table=sql.Identifier(self.base_table)
        )
        
        last_error = None
        attempt = 0
        
        while attempt <= retry_count:
            try:
                self.kb_search.cursor.execute(upsert_query, (path, json_data))
                result = self.kb_search.cursor.fetchone()
                
                # Commit the transaction
                self.kb_search.conn.commit()
                
                if result is not None:
                    if isinstance(result, dict):
                        returned_path = result['path']
                        was_inserted = result['was_inserted']
                    else:
                        returned_path, was_inserted = result
                    
                    operation = "inserted" if was_inserted else "updated"
                    return True, f"Successfully {operation} data for path: {returned_path}"
                else:
                    self.kb_search.conn.rollback()
                    raise Exception("Database operation completed but no result was returned")
                    
            except (OperationalError, InterfaceError) as e:
                # These are typically transient errors that may succeed on retry
                last_error = e
                
                try:
                    self.kb_search.conn.rollback()
                except Exception:
                    pass
                
                if attempt < retry_count:
                    time.sleep(retry_delay)
                    attempt += 1
                else:
                    break
                    
            except ValueError:
                # Don't retry on validation errors
                raise
                
            except Exception as e:
                # For other errors, attempt rollback but don't retry
                last_error = e
                
                try:
                    self.kb_search.conn.rollback()
                except Exception:
                    pass
                
                raise Exception(f"Error setting status data for path '{path}': {str(e)}")
        
        # If we've exhausted all retries
        error_msg = f"Failed to set status data for path '{path}' after {retry_count + 1} attempts"
        if last_error:
            error_msg += f": {str(last_error)}"
        
        raise Exception(error_msg)
    
    def set_multiple_status_data(self, path_data_pairs: Union[Dict[str, Dict], List[Tuple[str, Dict]]], 
                                retry_count: int = 3, retry_delay: float = 1.0) -> Tuple[bool, str, Dict]:
        """
        Performance-optimized method to update multiple path-data pairs in a single transaction.
        
        Args:
            path_data_pairs (dict or list): Dictionary of {path: data} or list of (path, data) tuples
            retry_count (int): Number of retries on failure (default: 3)
            retry_delay (float): Delay between retries in seconds (default: 1.0)
            
        Returns:
            tuple: (bool, str, dict) success flag, message, and results per path
        """
        if not path_data_pairs:
            raise ValueError("path_data_pairs cannot be empty")
        if retry_count < 0:
            raise ValueError("Retry count must be non-negative")
        if retry_delay < 0:
            raise ValueError("Retry delay must be non-negative")
        
        # Normalize input to list of tuples
        if isinstance(path_data_pairs, dict):
            pairs = list(path_data_pairs.items())
        else:
            pairs = list(path_data_pairs)
        
        # Validate all inputs first
        for path, data in pairs:
            if not path:
                raise ValueError(f"Path cannot be empty or None")
            if not isinstance(data, dict):
                raise ValueError(f"Data for path '{path}' must be a dictionary")
        
        # Pre-convert all data to JSON to avoid doing it multiple times on retries
        json_pairs = [(path, json.dumps(data, separators=(',', ':'))) for path, data in pairs]
        
        # Prepare the query with safe table name injection
        upsert_query = sql.SQL("""
            INSERT INTO {table} (path, data)
            VALUES (%s, %s)
            ON CONFLICT (path)
            DO UPDATE SET data = EXCLUDED.data
            RETURNING path, (xmax = 0) AS was_inserted
        """).format(
            table=sql.Identifier(self.base_table)
        )
        
        last_error = None
        attempt = 0
        
        while attempt <= retry_count:
            try:
                results = {}
                
                # Use batch operations within a transaction
                for path, json_data in json_pairs:
                    self.kb_search.cursor.execute(upsert_query, (path, json_data))
                    result = self.kb_search.cursor.fetchone()
                    
                    if result:
                        if isinstance(result, dict):
                            returned_path = result['path']
                            was_inserted = result['was_inserted']
                        else:
                            returned_path, was_inserted = result
                        
                        operation = "inserted" if was_inserted else "updated"
                        results[returned_path] = operation
                    else:
                        results[path] = "failed"
                
                # Commit all changes
                self.kb_search.conn.commit()
                
                success_count = len([r for r in results.values() if r != "failed"])
                return True, f"Successfully processed {success_count}/{len(pairs)} records", results
                
            except (OperationalError, InterfaceError) as e:
                # These are typically transient errors that may succeed on retry
                last_error = e
                
                try:
                    self.kb_search.conn.rollback()
                except Exception:
                    pass
                
                if attempt < retry_count:
                    time.sleep(retry_delay)
                    attempt += 1
                else:
                    break
                    
            except ValueError:
                # Don't retry on validation errors
                raise
                
            except Exception as e:
                # For other errors, attempt rollback but don't retry
                last_error = e
                
                try:
                    self.kb_search.conn.rollback()
                except Exception:
                    pass
                
                raise Exception(f"Error setting multiple status data: {str(e)}")
        
        # If we've exhausted all retries
        error_msg = f"Failed to set multiple status data after {retry_count + 1} attempts"
        if last_error:
            error_msg += f": {str(last_error)}"
        
        raise Exception(error_msg)
    
  

File: c_programs_and_containers/build_blocks/knowledge_base/postgres/data_structures/luajit/kb_status_data.lua
--[[
  KB_Status_Data - Status data read/write for the knowledge base.
  LuaJIT port using luadbi-postgresql and dkjson.

  Usage:
    local KB_Search      = require("kb_search")
    local KB_Status_Data = require("kb_status_data")
    local kb     = KB_Search.new({ ... })
    local status = KB_Status_Data.new(kb, "my_database")
]]

local dkjson = require("dkjson")

local KB_Status_Data = {}
KB_Status_Data.__index = KB_Status_Data

function KB_Status_Data.new(kb_search, database)
  local self = setmetatable({}, KB_Status_Data)
  self.kb_search  = kb_search
  self.dbh        = kb_search:get_connection()
  self.base_table = database .. "_status"
  return self
end

---------------------------------------------------------------------------
-- Helpers
---------------------------------------------------------------------------

local function esc(val)
  if val == nil then return "NULL" end
  local s = tostring(val)
  s = s:gsub("'", "''")
  return "'" .. s .. "'"
end

local function sleep(sec)
  local ok, socket = pcall(require, "socket")
  if ok then socket.sleep(sec) else
    local t = os.clock() + sec
    while os.clock() < t do end
  end
end

---------------------------------------------------------------------------
-- Node discovery
---------------------------------------------------------------------------

function KB_Status_Data:find_node_id(kb, node_name, properties, node_path)
  local results = self:find_node_ids(kb, node_name, properties, node_path)
  if #results == 0 then
    error(string.format("No node found: kb=%s, name=%s", tostring(kb), tostring(node_name)))
  end
  if #results > 1 then
    error(string.format("Multiple nodes (%d) found: kb=%s, name=%s",
      #results, tostring(kb), tostring(node_name)))
  end
  return results[1]
end

function KB_Status_Data:find_node_ids(kb, node_name, properties, node_path)
  self.kb_search:clear_filters()
  self.kb_search:search_label("KB_STATUS_FIELD")

  if kb        then self.kb_search:search_kb(kb) end
  if node_name then self.kb_search:search_name(node_name) end
  if properties and type(properties) == "table" then
    for k, v in pairs(properties) do
      self.kb_search:search_property_value(k, v)
    end
  end
  if node_path then self.kb_search:search_path(node_path) end

  local rows = self.kb_search:execute_query()
  if not rows or #rows == 0 then
    error(string.format("No nodes found: kb=%s, name=%s", tostring(kb), tostring(node_name)))
  end
  return rows
end

---------------------------------------------------------------------------
-- Status data CRUD
---------------------------------------------------------------------------

function KB_Status_Data:get_status_data(path)
  if not path or path == "" then error("Path cannot be empty or nil") end

  local sql_str = string.format(
    "SELECT data, path FROM %s WHERE path = %s LIMIT 1",
    self.base_table, esc(path))
  local row = self.kb_search:_raw_query_one(sql_str)

  if not row then
    error(string.format("No data found for path: %s", path))
  end

  local data = row.data
  if type(data) == "string" then
    local decoded, _, err = dkjson.decode(data)
    if not decoded then
      error(string.format("Failed to decode JSON data for path '%s': %s", path, tostring(err)))
    end
    data = decoded
  end

  return data, row.path
end

function KB_Status_Data:get_multiple_status_data(paths)
  if not paths or #paths == 0 then return {} end
  if type(paths) ~= "table" then paths = { paths } end

  local placeholders = {}
  for i, p in ipairs(paths) do
    placeholders[i] = esc(p)
  end

  local sql_str = string.format(
    "SELECT data, path FROM %s WHERE path IN (%s)",
    self.base_table, table.concat(placeholders, ","))

  local res = self.kb_search:_raw_query(sql_str)
  local rv = {}
  for _, row in ipairs(res) do
    local data = row.data
    if type(data) == "string" then
      data = dkjson.decode(data) or data
    end
    rv[row.path] = data
  end
  return rv
end

function KB_Status_Data:set_status_data(path, data, retry_count, retry_delay)
  retry_count = retry_count or 3
  retry_delay = retry_delay or 1.0

  if not path or path == "" then error("Path cannot be empty or nil") end
  if type(data) ~= "table" then error("Data must be a table") end

  local json_data = dkjson.encode(data)

  local upsert_sql = string.format([[
    INSERT INTO %s (path, data) VALUES (%s, %s::jsonb)
    ON CONFLICT (path)
    DO UPDATE SET data = EXCLUDED.data
    RETURNING path, (xmax = 0) AS was_inserted
  ]], self.base_table, esc(path), esc(json_data))

  local last_err
  for attempt = 0, retry_count do
    local ok, result = pcall(function()
      local row = self.kb_search:_raw_query_one(upsert_sql)
      self.kb_search:commit()
      if not row then
        self.kb_search:rollback()
        error("Database operation completed but no result was returned")
      end
      return row
    end)

    if ok then
      local row = result
      local was_ins = (row.was_inserted == true or row.was_inserted == "t")
      local op = was_ins and "inserted" or "updated"
      return true, string.format("Successfully %s data for path: %s", op, row.path)
    end

    last_err = result
    self.kb_search:rollback()

    if attempt < retry_count then sleep(retry_delay) end
  end

  error(string.format("Failed to set status data for path '%s' after %d attempts: %s",
    path, retry_count + 1, tostring(last_err)))
end

function KB_Status_Data:set_multiple_status_data(path_data_pairs, retry_count, retry_delay)
  retry_count = retry_count or 3
  retry_delay = retry_delay or 1.0

  if not path_data_pairs then error("path_data_pairs cannot be empty") end

  -- Normalise to array of {path, json_string}
  local pairs_list = {}
  if path_data_pairs[1] and type(path_data_pairs[1]) == "table" then
    -- array of {path, data}
    for _, pair in ipairs(path_data_pairs) do
      if not pair[1] or pair[1] == "" then error("Path cannot be empty") end
      if type(pair[2]) ~= "table" then error("Data must be a table") end
      pairs_list[#pairs_list + 1] = { pair[1], dkjson.encode(pair[2]) }
    end
  else
    -- dict-style
    for p, d in pairs(path_data_pairs) do
      if type(p) ~= "string" or p == "" then error("Path cannot be empty") end
      if type(d) ~= "table" then error("Data must be a table") end
      pairs_list[#pairs_list + 1] = { p, dkjson.encode(d) }
    end
  end

  local last_err
  for attempt = 0, retry_count do
    local ok, err_or_results = pcall(function()
      local results = {}
      for _, pair in ipairs(pairs_list) do
        local upsert_sql = string.format([[
          INSERT INTO %s (path, data) VALUES (%s, %s::jsonb)
          ON CONFLICT (path)
          DO UPDATE SET data = EXCLUDED.data
          RETURNING path, (xmax = 0) AS was_inserted
        ]], self.base_table, esc(pair[1]), esc(pair[2]))

        local row = self.kb_search:_raw_query_one(upsert_sql)
        if row then
          local was_ins = (row.was_inserted == true or row.was_inserted == "t")
          results[row.path] = was_ins and "inserted" or "updated"
        else
          results[pair[1]] = "failed"
        end
      end
      self.kb_search:commit()
      return results
    end)

    if ok then
      local results = err_or_results
      local success_count = 0
      for _, v in pairs(results) do
        if v ~= "failed" then success_count = success_count + 1 end
      end
      return true,
        string.format("Successfully processed %d/%d records", success_count, #pairs_list),
        results
    end

    last_err = err_or_results
    self.kb_search:rollback()
    if attempt < retry_count then sleep(retry_delay) end
  end

  error(string.format("Failed to set multiple status data after %d attempts: %s",
    retry_count + 1, tostring(last_err)))
end

return KB_Status_Data

